README.md
# Attention based DisOrder PredicTor

This repository containes the code and the trained models for **intrinsic protein disorder prediction through deep bidirectional transformers** from Peptone Ltd.

[![DOI](https://zenodo.org/badge/426679803.svg)](https://zenodo.org/badge/latestdoi/426679803)
[![GitHub Super-Linter](https://github.com/peptoneltd/ADOPT/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/marketplace/actions/super-linter)

ADOPT has been introduced in our paper [ADOPT: intrinsic protein disorder prediction through deep bidirectional transformers](https://www.biorxiv.org/content/10.1101/2022.05.25.493416v3) and it's also available as webserver at [adopt.peptone.io](https://adopt.peptone.io).

<img src="https://github.com/PeptoneLtd/ADOPT/blob/main/media/adopt_attention.gif" width="600"/>

Our disorder predictor is made up of two main blocks, namely: a **self-supervised encoder** and a **supervised disorder predictor**. We use [Facebook’s Evolutionary Scale Modeling (ESM)](https://github.com/facebookresearch/esm) library to extract dense residue evel representations, which feed the  supervised machine learning based predictor.

The ESM library exploits a set of deep Transformer encoder models, which processes character sequences of amino acids as inputs.

ADOPT makes use of two datasets: the [CheZoD  “1325” and the CheZoD “117”](https://github.com/protein-nmr/CheZOD) databases containing 1325 and 117 sequences, respectively, together with their  residue level **Z-scores**.

## Table of Contents

- [Attention based DisOrder PredicTor](#attention-based-disorder-predictor)
  - [Table of Contents](#table-of-contents)
  - [Intrinsic disorder trained models](#intrinsic-disorder-trained-models)
  - [Usage](#usage)
    - [Quick start](#quick-start)
    - [MSA setting (optional)](#msa-setting-optional)
    - [Scripts](#scripts)
    - [Notebooks](#notebooks)
    - [Compute residue level representations](#compute-residue-level-representations)
    - [Predict intrinsic disorder with ADOPT](#predict-intrinsic-disorder-with-adopt)
    - [Train ADOPT disorder predictor](#train-adopt-disorder-predictor)
    - [Run benchmarks](#run-benchmarks)
  - [Citations](#citations)
  - [Licence](#licence)

## Intrinsic disorder trained models

| Model | Pre-trained model | Datasets | Split level | CV |
|-------|-------------------|----------|-------------|----|
| `lasso_esm-1b_cleared_residue` | ESM-1b | **Chezod 1325 cleared** and **Chezod 117** | residue | :x: |
| `lasso_esm-1v_cleared_residue` | ESM-1v | **Chezod 1325 cleared** and **Chezod 117** | residue | :x: |
| `lasso_esm-msa_cleared_residue` | ESM-MSA | **Chezod 1325 cleared** and **Chezod 117** | residue | :x: |
| `lasso_combined_cleared_residue` | Combined | **Chezod 1325 cleared** and **Chezod 117** | residue | :x: |
| `lasso_esm-1b_residue_cv` | ESM-1b | **Chezod 1325** | residue | :white_check_mark: |
| `lasso_esm-1v_residue_cv` | ESM-1v | **Chezod 1325** | residue | :white_check_mark: |
| `lasso_esm-msa_residue_cv` | ESM-MSA | **Chezod 1325** | residue | :white_check_mark: |
| `lasso_esm-1b_cleared_residue_cv` | ESM-1b | **Chezod 1325 cleared** | residue | :white_check_mark: |
| `lasso_esm-1v_cleared_residue_cv` | ESM-1v | **Chezod 1325 cleared** | residue | :white_check_mark: |
| `lasso_esm-msa_cleared_residue_cv` | ESM-MSA | **Chezod 1325 cleared** | residue | :white_check_mark: |
| `lasso_esm-1b_cleared_sequence_cv` | ESM-1b | **Chezod 1325 cleared** | sequence | :white_check_mark: |
| `lasso_esm-1v_cleared_sequence_cv` | ESM-1v | **Chezod 1325 cleared** | sequence | :white_check_mark: |
| `lasso_esm-msa_cleared_sequence_cv` | ESM-MSA | **Chezod 1325 cleared** | sequence | :white_check_mark: |

## Usage

### Quick start

Prerequisites (we suggest creating a dedicated python venv or conda env)

```
pip install \
  pandas \
  fair-esm \
  biopython \
  bertviz \
  skl2onnx \
  onnxruntime \
  spacy \
  plotly \
  wandb 
```

Install the **adopt** package:

Run

```bash
git clone https://github.com/PeptoneInc/ADOPT.git
cd ADOPT
git submodule update --init --recursive
python setup.py install
```

Then, you can predict the intrinsic disorder of each reesidue in a protein sequence, as follows:

```python
from adopt import MultiHead, ZScorePred

# Prepare protein sequence and name i.e brmid
SEQUENCE = "SLQDGVRQSRASDKQTLLPNDQLYQPLKDREDDQYSHLQGNQLRRN"
PROTID = "Protein 18890"

# Choose model type and training strategy
MODEL_TYPE = "esm-1b"
STRATEGY = "train_on_cleared_1325_test_on_117_residue_split"

# Extract residue level representations
multi_head = MultiHead(MODEL_TYPE)
representation, tokens = multi_head.get_representation(SEQUENCE, PROTID)

# Predict the Z score related to each residue in the sequence specified above
z_score_pred = ZScorePred(STRATEGY, MODEL_TYPE)
predicted_z_scores = z_score_pred.get_z_score(representation)
````

### MSA setting (optional)

In order to enable the ```esm-msa``` based variant of ADOPT, MSAs for each sequence are also required.
We provide a stand alone, ```docker``` based tool you must use to exploit all the functionalities of ADOPT for `msa` related tasks.

#### First time setup

As a prerequisite, you must have [Docker](https://www.docker.com/) installed.

Clone the ADOPT repository, go to the ADOPT directory and run the [MSA scripts](#scripts) you are interested in.

#### Notes

The `$LOCAL_MSA_DIR` in the **MSA scripts** serves as the main directory for the MSA related procedures and can be empty
initially when running the above scripts. Under the hood, each **MSA script** will:

 1. Download [uniclust](http://gwdu111.gwdg.de/~compbiol/uniclust/) dataset (in this case "2020.06") into
  the ```$LOCAL_MSA_DIR/databases``` subdirectory. \
  **!NOTE**: under the hood,
  [ADOPT](scripts/uniclust_download.py) checks, whether uniclust is already in this subdirectory. If not, downloading
  can take several hours, given the size of this dataset is approx 180GB! Download step is skipped only if the ```$LOCAL_MSA_DIR/databases```
  folder is non empty and the tar file (<cite>UniRef30_2020_06_hhsuite.tar.gz</cite>) is found in the ```$LOCAL_MSA_DIR``` folder.

 2. Once the relevant uniclust is there, a docker image named ```msa-gen-adopt``` is run with the volume ```$LOCAL_MSA_DIR```
mounted on it.

Note that this setup procedure creates four subfolders:

<pre>
+-- $LOCAL_MSA_DIR
|   +-- databases
|   +-- msas
|   +-- msa_fastas
|   +-- esm_msa_reprs
</pre>

```databases``` will hold the [uniclust](http://gwdu111.gwdg.de/~compbiol/uniclust/);
```msas``` is where MSAs (```.a3m``` files) will be saved later, see STEP 2 below;
```msa_fastas``` is where ```.fasta``` files already used for MSA queries will be saved;
```esm_msa_reprs``` is allocated for potential ```esm-msa``` representations;

The MSAs will be placed in the ```$LOCAL_MSA_DIR/msas``` folder.

### More notes

You can set the ```ESM_MODELS_DIR``` and ```ADOPT_MODELS_DIR``` respectively to paths where the [ESM](https://dl.fbaipublicfiles.com/fair-esm/models/esm1v_t33_650M_UR90S_1.pt) and [ADOPT](https://adopt-models.s3.eu-west-2.amazonaws.com/models.zip) pretrained models are stored.
All models will be downloaded from public repositories if not found locally.

### Scripts

The [scripts](scripts) directory contains:

- [inference](scripts/adopt_inference.sh) script to predict, in bulk, the disorder of each residue in each protein sequence reported in a FASTA file, with ADOPT where you need to specify:
  - `NEW_PROT_FASTA_FILE_PATH` defining your FASTA file path
  - `NEW_PROT_RES_REPR_DIR_PATH` defining where the residue level representations will be extracted
- [training](scripts/adopt_chezod_training.sh) script to train the ADOPT where you need to specify:
  - `TRAIN_STRATEGY` defining the training strategy you want to use
- [MSA inference](scripts/adopt_msa_inference.sh) script, which allows to perform [inference](scripts/adopt_inference.sh) also with the `esm-msa` model. The predicted Z scores will be written on the host (**optional**)
- [MSA training](scripts/adopt_chezod_msa_training.sh) script, which allows to perform [training](scripts/adopt_chezod_training.sh) also with the `esm-msa` model. The trained models will be written in the `ADOPT/models` directory (**optional**)

### Notebooks

The [notebooks](notebooks) directory contains:

- [disorder prediction](notebooks/adopt_disorder_prediction.ipynb) notebook
- [multi-head attention weights visualisation](notebooks/adopt_attention_viz.ipynb) notebook

### Compute residue level representations

In order to predict the **Z score** related to each residue in a protein sequence, we have to compute the residue level representations, extracted from the pretrained model.

In the ADOPT directory run:

```bash
python adopt/embedding.py <fasta_file_path> \
                          <residue_level_representation_dir>
```

Where:

- `<fasta_file_path>` defines the FASTA file containing the proteins for which you want to compute the intrinsic disorder
- `<residue_level_representation_dir>` defines the path where you want to save the residue level representations
- `--msa` runs the [MSA procedure](#msa-setting-optional) to get `esm-msa` representations. We suggest you take a look to the [MSA inference](scripts/adopt_msa_inference.sh) script as a quick example (**optional**)
- `-h` shows help message and exit

A subdirectory containing the residue level representation extracted from each pre-trained model available will be created under both the `residue_level_representation_dir`.

Important to note that in order to obtain the representations from the ```esm-msa``` model as well, the relevant MSAs have to
be placed in the root directory `/msas` in the system, where ADOPT is running. The MSAs can be created as described in
the MSA setting above.

### Predict intrinsic disorder with ADOPT

Once we have extracted the residue level representations we can predict the intrinsic disorder (Z score).

In the ADOPT directory run:

```bash
python adopt/inference.py <inference_fasta_file> \
                          <inference_repr_dir> \
                          <predicted_z_scores_file> \
                          --train_strategy <training_strategy> \
                          --model_type <model_type> 
```

Where:

- `<inference_fasta_file>` defines the FASTA file containing the proteins for which you want to compute the intrinsic disorder
- `<inference_repr_dir>` defines the path where you've already saved the residue level representations
- `<predicted_z_scores_file>` defines the path where you want the Z scores to be saved
- `--train_strategy` defines the **training strategies** defined below
- `--model_type` defines the pre-trained model we want to use. We suggest you use the `esm-1b` model
- `-h` shows help message and exit

The output is a `.json` file contains the Z scores related to each residue of each protein in the FASTA file where you put the proteins you are intereseted in.

| Training strategy | Pre-trained models |
|-------------------|-------------------|
| `train_on_cleared_1325_test_on_117_residue_split` | `esm-1b`, `esm-1v`, `esm-msa` and `combined` |
| `train_on_1325_cv_residue_split`| `esm-1b`, `esm-1v` and `esm-msa` |
| `train_on_cleared_1325_cv_residue_split`| `esm-1b`, `esm-1v` and `esm-msa` |
| `train_on_cleared_1325_cv_sequence_split`| `esm-1b`, `esm-1v` and `esm-msa` |
| `train_on_total`| `esm-1b`, `esm-1v`|

### Train ADOPT disorder predictor

Once we have extracted the residue level representations of the protein for which we want to predict the intrinsic disorder (Z score), we can train the predictor.

**NOTE**: This step is not mandatory because we've already trained such models. You can find them in the [models](https://adopt-models.s3.eu-west-2.amazonaws.com/models.zip) bucket.

In the ADOPT directory run:

```bash
python adopt/training.py <train_json_file_path> \
                         <test_json_file_path> \
                         <train_residue_level_representation_dir> \
                         <test_residue_level_representation_dir> \
                         --train_strategy <training_strategy> 
```

Where:

- `<train_json_file_path>` defines the JSON containing the proteins we want to use as *training set*
- `<test_json_file_path>` defines the JSON containing the proteins we want to use as *test set*
- `<train_residue_level_representation_dir>` defines the path where we saved the residue level representations of the proteins in the *training set*
- `<test_residue_level_representation_dir>` defines the path where we saved the residue level representations of the proteins in the *test set*
- `--train_strategy` defines the **training strategies** defined above
- `--msa` runs the [MSA procedure](msa-setting-optional) to get trained models fed with the `esm-msa` representations. We suggest you take a look to the [MSA training](scripts/adopt_chezod_msa_training.sh) script as a quick example (**optional**)
- `-h` shows help message and exit

### Run benchmarks

Once we have extracted the residue level representations we can benchmark ADOPT against other methods.

In the ADOPT directory run:

```bash
python adopt/benchmarks.py <benchmark_data_path> \
                           <train_json_file_path> \
                           <test_json_file_path> \
                           <train_residue_level_representation_dir> \
                           <test_residue_level_representation_dir> \
                           --train_strategy <training_strategy> 
```

Where:

- `<benchmark_data_path>` defines the directory containing the predictions of the method we want to benchmark againbst ADOPT
- `-h` shows help message and exit

#### AlphaFold2 benchmarks (optional)

We benchmarked ADOPT against [AlphaFold2](https://github.com/deepmind/alphafold) computing the spearman correlations between actual Z-scores and predicted pLDDT<sub>5</sub> scores along with actual Z-scores and predicted SASA<sub>5</sub> scores,
obtained by AlphaFold2, collected for the task linked to the model evaluated on the [CheZoD “117”](datasets/chezod_117_all.fasta) validation set and described in the ADOPT [paper](https://link_to_paper).

As a prerequisite, you must have [Docker](https://www.docker.com/) installed.

Run:

```bash
docker run ghcr.io/peptoneinc/adopt_alphafold2_comparison:1.0.2
```

[Here](scripts/get_alphafold2_correlations.py) is the script used to extract the correlations and [here](https://alphafold2-chezod-predictions.s3.eu-central-1.amazonaws.com/data.tar.gz) are the predictions obtained from Alphafold2.

## Citations <a name="citations"></a>

If you use this work in your research, please cite the the relevant paper:

```bibtex
@article {Tamiola2022.05.25.493416,
  author = {Tamiola, Kamil and Fisicaro, Carlo and Dutton, Oliver and Hoffmann, Falk and Henderson, Louie and Owens, Benjamin M.J. and Heberling, Matthew  and Redl, Istvan},
  title = {ADOPT: intrinsic protein disorder prediction through deep bidirectional transformers},
  elocation-id = {2022.05.25.493416},
  year = {2022},
  doi = {10.1101/2022.05.25.493416},
  publisher = {Cold Spring Harbor Laboratory},
  abstract = {Intrinsically disordered proteins (IDP) are important in a broad range of biological functions and are involved in many diseases. An understanding of intrinsic disorder is key to develop drugs against IDPs. Experimental characterization of IDPs are expensive and less efficient and
  demand the development of computational tools. Here, we present ADOPT, a new predictor of protein disorder. ADOPT is a deep bidirectional transformer, which extracts dense residue level representations from Facebook{\textquoteright}s Evolutionary Scale Modeling (ESM) library. Using the experimentally designed CheZod database as a training and test dataset for protein disorder, it predicts Z scores and protein disorder with new state-of-the-art
  performance in a few seconds. We show that ADOPT offers substantial improvement in comparison to previous predictors with a Spearman correlation coefficient between experimental and computational Z scores of 0.69. We identify the coordinates which are relevant for the prediction performance and show that good performance can already gained with less than 100 features. We believe that ADOPT will be a useful tool for all experimental scientists 
  working with intrinsically disordered proteins. It is available as a standalone package at https://github.com/PeptoneInc/ADOPT.git.Competing Interest StatementThe authors have declared no competing interest.},
  URL = {https://www.biorxiv.org/content/early/2022/05/26/2022.05.25.493416},
  eprint = {https://www.biorxiv.org/content/early/2022/05/26/2022.05.25.493416.full.pdf},
  journal = {bioRxiv}
}
```

## Licence

This source code is licensed under the MIT license found in the `LICENSE` file in the root directory of this source tree.


CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/peptoneltd/ADOPT/tree/HEAD)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.5.0...HEAD)

**Closed issues:**

- module 'esm' has no attribute 'pretrained' [\#40](https://github.com/PeptoneLtd/ADOPT/issues/40)

## [v0.5.0](https://github.com/peptoneltd/ADOPT/tree/v0.5.0) (2023-03-01)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.5...v0.5.0)

## [v0.4.5](https://github.com/peptoneltd/ADOPT/tree/v0.4.5) (2022-11-15)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.4...v0.4.5)

**Closed issues:**

- Nonexistant folder in dockerfile [\#38](https://github.com/PeptoneLtd/ADOPT/issues/38)

**Merged pull requests:**

- Dev [\#39](https://github.com/PeptoneLtd/ADOPT/pull/39) ([CFisicaro](https://github.com/CFisicaro))

## [v0.4.4](https://github.com/peptoneltd/ADOPT/tree/v0.4.4) (2022-11-02)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.3...v0.4.4)

**Implemented enhancements:**

- Add default options [\#33](https://github.com/PeptoneLtd/ADOPT/issues/33)

**Merged pull requests:**

- Refactor code + simple webservice [\#36](https://github.com/PeptoneLtd/ADOPT/pull/36) ([fabio-peptone](https://github.com/fabio-peptone))

## [v0.4.3](https://github.com/peptoneltd/ADOPT/tree/v0.4.3) (2022-05-27)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.2...v0.4.3)

**Closed issues:**

- Use ESM as submodule once ADOPT goes public [\#3](https://github.com/PeptoneLtd/ADOPT/issues/3)

**Merged pull requests:**

- Dev [\#35](https://github.com/PeptoneLtd/ADOPT/pull/35) ([CFisicaro](https://github.com/CFisicaro))

## [v0.4.2](https://github.com/peptoneltd/ADOPT/tree/v0.4.2) (2022-01-26)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.1...v0.4.2)

**Implemented enhancements:**

- @falkph please remove all the installation commands [\#30](https://github.com/PeptoneLtd/ADOPT/issues/30)

**Merged pull requests:**

- Adopt example [\#29](https://github.com/PeptoneLtd/ADOPT/pull/29) ([falkph](https://github.com/falkph))
- Viz routines [\#25](https://github.com/PeptoneLtd/ADOPT/pull/25) ([CFisicaro](https://github.com/CFisicaro))

## [v0.4.1](https://github.com/peptoneltd/ADOPT/tree/v0.4.1) (2021-12-09)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.4.0...v0.4.1)

## [v0.4.0](https://github.com/peptoneltd/ADOPT/tree/v0.4.0) (2021-12-09)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.3.0...v0.4.0)

**Closed issues:**

- Include ESM-MSA pre trained model [\#2](https://github.com/PeptoneLtd/ADOPT/issues/2)

**Merged pull requests:**

- Esm msa dev [\#21](https://github.com/PeptoneLtd/ADOPT/pull/21) ([CFisicaro](https://github.com/CFisicaro))

## [v0.3.0](https://github.com/peptoneltd/ADOPT/tree/v0.3.0) (2021-11-22)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.2.1...v0.3.0)

**Closed issues:**

- Add notebooks description in README [\#18](https://github.com/PeptoneLtd/ADOPT/issues/18)
- Add scripts usage in README [\#16](https://github.com/PeptoneLtd/ADOPT/issues/16)
- Use argparse library in place of getopt library [\#9](https://github.com/PeptoneLtd/ADOPT/issues/9)

**Merged pull requests:**

- Develop [\#20](https://github.com/PeptoneLtd/ADOPT/pull/20) ([CFisicaro](https://github.com/CFisicaro))

## [v0.2.1](https://github.com/peptoneltd/ADOPT/tree/v0.2.1) (2021-11-16)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.2.0...v0.2.1)

**Closed issues:**

- Add jupyter notebooks tutorial [\#5](https://github.com/PeptoneLtd/ADOPT/issues/5)

**Merged pull requests:**

- Interactive disorder pred [\#19](https://github.com/PeptoneLtd/ADOPT/pull/19) ([CFisicaro](https://github.com/CFisicaro))

## [v0.2.0](https://github.com/peptoneltd/ADOPT/tree/v0.2.0) (2021-11-15)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.1.1...v0.2.0)

**Merged pull requests:**

- Attention viz [\#17](https://github.com/PeptoneLtd/ADOPT/pull/17) ([CFisicaro](https://github.com/CFisicaro))

## [v0.1.1](https://github.com/peptoneltd/ADOPT/tree/v0.1.1) (2021-11-15)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.1.0...v0.1.1)

**Implemented enhancements:**

- Onnx runtime [\#15](https://github.com/PeptoneLtd/ADOPT/pull/15) ([CFisicaro](https://github.com/CFisicaro))

**Fixed bugs:**

- Bad inference results [\#12](https://github.com/PeptoneLtd/ADOPT/issues/12)

## [v0.1.0](https://github.com/peptoneltd/ADOPT/tree/v0.1.0) (2021-11-13)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/v0.0.1...v0.1.0)

**Closed issues:**

- Add bash script for inference [\#8](https://github.com/PeptoneLtd/ADOPT/issues/8)
- Add bash script for training [\#7](https://github.com/PeptoneLtd/ADOPT/issues/7)
- Add CHANGELOG [\#6](https://github.com/PeptoneLtd/ADOPT/issues/6)

**Merged pull requests:**

- Super linter integration [\#11](https://github.com/PeptoneLtd/ADOPT/pull/11) ([CFisicaro](https://github.com/CFisicaro))

## [v0.0.1](https://github.com/peptoneltd/ADOPT/tree/v0.0.1) (2021-11-11)

[Full Changelog](https://github.com/peptoneltd/ADOPT/compare/a1bb9c7ed570348da5f39b1fd7ed6df6608236ab...v0.0.1)

**Closed issues:**

- Better README [\#1](https://github.com/PeptoneLtd/ADOPT/issues/1)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


adopt/utils.py
# Copyright (c) 2021 Peptone.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import esm
import numpy as np
import onnxruntime as rt
import pandas as pd
import torch
from Bio import SeqIO
from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import FloatTensorType
from sklearn import linear_model
import os

from adopt import constants

"@generated"
# throw away the missing values, if the drop_missing flag is set to True, i.e. where z-scores are  999
def pedestrian_input(indexes, df, path, z_col="z-score", msa=False, drop_missing=True):
    zeds = []
    exes = []

    for k in range(len(indexes)):
        if not msa:
            repr_esm = (
                torch.load(f"{path}{indexes[k]}.pt")["representations"][33]
                .clone()
                .cpu()
                .detach()
            )
        else:
            repr_esm = (
                torch.load(f"{path}{indexes[k]}.pt")["representations"]
                .clone()
                .cpu()
                .detach()
            )
        z_s = np.array(df[df["brmid"] == indexes[k]][z_col].to_numpy()[0])
        if drop_missing:
            idxs = np.where(z_s != 999)[0]
        else:
            idxs = np.arange(len(z_s))

        for i in idxs:
            zeds.append(z_s[i])
            exes.append(repr_esm[i].numpy())
    return np.array(exes), np.array(zeds)


# collect the path to representations according to model type and train vs test set
def representation_path(path_chezod_1325_repr, path_chezod_117_repr, msa):
    repr_path = {}
    if msa:
        model_types = constants.msa_model_types
    else:
        model_types = constants.model_types
    for model_type in model_types:
        repr_path[model_type] = {
            "1325": str(path_chezod_1325_repr) + "/" + model_type + "/",
            "117": str(path_chezod_117_repr) + "/" + model_type + "/",
        }
    return repr_path


def df_to_fasta(df, fasta_out_path):
    ofile = open(fasta_out_path, "w")
    for index, row in df.iterrows():
        ofile.write(">" + row["brmid"] + "\n" + row["sequence"] + "\n")
    ofile.close()


def fasta_to_df(fasta_input):
    with open(fasta_input) as fasta_file:
        identifiers = []
        sequences = []
        for seq_record in SeqIO.parse(fasta_file, "fasta"):  # (generator)
            identifiers.append(seq_record.id)
            sequences.append("".join(seq_record.seq))

    df = pd.DataFrame({"brmid": identifiers, "sequence": sequences})
    return df


def save_onnx_model(columns_shape, reg, model_name):
    initial_type = [("float_input", FloatTensorType([None, columns_shape]))]
    onx = convert_sklearn(reg, initial_types=initial_type)
    with open(model_name, "wb") as f:
        f.write(onx.SerializeToString())


def get_onnx_model_preds(model_name, input_data):
    sess = rt.InferenceSession(model_name)
    input_name = sess.get_inputs()[0].name
    label_name = sess.get_outputs()[0].name
    pred_onx = sess.run([label_name], {input_name: input_data})[0]
    return pred_onx


def get_esm_output(model, alphabet, data, msa):
    if msa:
        msa_transformer = model.eval()
        msa_batch_converter = alphabet.get_batch_converter()
        _, _, msa_batch_tokens = msa_batch_converter(data)
        with torch.no_grad():
            results = msa_transformer(msa_batch_tokens, repr_layers=[12], return_contacts=True)
    else:
        batch_converter = alphabet.get_batch_converter()
        batch_labels, batch_strs, batch_tokens = batch_converter(data)
        # Extract per-residue representations (on CPU)
        with torch.no_grad():
            results = model(batch_tokens, repr_layers=[33], return_contacts=True)
    return results


def load_model_and_alphabet(model_name):
    models_dir = os.environ.get("ESM_MODELS_DIR")
    try: # try to find model locally
        path = os.path.join(models_dir, f"{model_name}.pt")
        print(f"Trying to load ESM model from {path}")
        return esm.pretrained.load_model_and_alphabet_local(path)
    #TODO: catch appropriate exception 
    except Exception as e:
        print(type(e))    # the exception instance
        print(e.args)     # arguments stored in .args
        print(e)
        print("ESM model not found locally, downloading from torchub")
        return esm.pretrained.load_model_and_alphabet_hub(model_name)


def get_model_and_alphabet(model_type, data):
    # Load ESM model
    if model_type == "esm-1b":
        model, alphabet = load_model_and_alphabet("esm1b_t33_650M_UR50S")
        results = get_esm_output(model, alphabet, data, False)
    elif model_type == "esm-1v":
        model, alphabet = load_model_and_alphabet("esm1v_t33_650M_UR90S_1")
        results = get_esm_output(model, alphabet, data, False)
    elif model_type == 'esm-msa':
        model, alphabet = load_model_and_alphabet("esm_msa1b_t12_100M_UR50S")
        results = get_esm_output(model, alphabet, data, True)
    else:
        model_esm1b, alphabet_esm1b = load_model_and_alphabet("esm1b_t33_650M_UR50S")
        model_esm1v, alphabet_esm1v = load_model_and_alphabet("esm1v_t33_650M_UR90S_1")
        results_esm1b = get_esm_output(model_esm1b, alphabet_esm1b, data)
        results_esm1v = get_esm_output(model_esm1v, alphabet_esm1v, data)
        results = [results_esm1b, results_esm1v]
    return results


def get_model_alphabet_msa(model_type):
    models = []
    alphabets = []
    msa = []

    if model_type == "esm-1b":
        model, alphabet = load_model_and_alphabet("esm1b_t33_650M_UR50S")
        return [model], [alphabet], [False]
    
    if model_type == "esm-1v":
        model, alphabet = load_model_and_alphabet("esm1v_t33_650M_UR90S_1")
        return [model], [alphabet], [False]
    
    if model_type == 'esm-msa':
        model, alphabet = load_model_and_alphabet("esm_msa1b_t12_100M_UR50S")
        return [model], [alphabet], [True]
    
    for s in ("esm1b_t33_650M_UR50S", "esm1v_t33_650M_UR90S_1"):
        m, a = load_model_and_alphabet(s)
        models.append(m)
        alphabets.append(a)
        msa.append(False)
        
    return models, alphabets, msa
    

def get_residue_class(predicted_z_scores):
    residues_state = []
    for n, zscore in enumerate(predicted_z_scores):
        residues_dict = {}
        if zscore < 3:
            residues_dict["label"] = constants.structure_dict["Fully disordered"]
        elif 3 <= zscore < 8:
            residues_dict["label"] = constants.structure_dict["Partially disordered"]
        elif zscore >= 11:
            residues_dict["label"] = constants.structure_dict["Structured"]
        else:
            residues_dict["label"] = constants.structure_dict["Flexible loops"]

        residues_dict["start"] = n
        residues_dict["end"] = n + 1
        residues_state.append(residues_dict)
    return residues_state


def stability_selection_prob(ex_train, zed_train, model_picked, sample_size, reg_param):
    sample_idxs = np.random.choice(np.arange(ex_train[model_picked].shape[0]), sample_size, replace=False) 
    ex_train_filtered = np.take(ex_train[model_picked], sample_idxs, axis=0)
    zed_train_filtered = np.take(zed_train[model_picked], sample_idxs, axis=0)
    # Lasso regression
    reg = linear_model.Lasso(alpha=reg_param, max_iter=10000)
    reg.fit(ex_train_filtered, zed_train_filtered)
    reg_coef = abs(reg.coef_)
    return reg_coef

  
def get_esm_models(msa):
    if msa:
        models = constants.esm_msa_models
    else:
        models = constants.esm_models
    return models


def get_model_types(msa):
    if msa:
        model_types = constants.msa_model_types
    else:
        model_types = constants.model_types
    return model_types
